File: tool/download_moneroc_prebuilds.dart
import 'dart:io';

import 'package:dio/dio.dart';
import 'package:archive/archive_io.dart';

final _dio = Dio();

final List<String> triplets = [
  "x86_64-linux-gnu", // linux desktop - majority of users onlinux
  // "i686-linux-gnu", // not supported by cake
  // "i686-meego-linux-gnu", // sailfishos (emulator)- not supported by cake
  // "aarch64-linux-gnu", // not (yet) supported by cake - (mostly) mobile linux
  // "aarch64-meego-linux-gnu", // sailfishos - not supported by cake
  "x86_64-linux-android",
  // "i686-linux-android", // not supported by monero_c - mostly old android emulators
  "aarch64-linux-android",
  "armv7a-linux-androideabi",
  // "i686-w64-mingw32", // 32bit windows - not supported by monero_c
  "x86_64-w64-mingw32",
  // "x86_64-apple-darwin11", // Intel macbooks (contrib) - not used by cake
  // "aarch64-apple-darwin11", // apple silicon macbooks (contrib) - not used by cake
  // "x86_64-host-apple-darwin", // not available on CI (yet)
  "aarch64-apple-darwin", // apple silicon macbooks
  "x86_64-apple-darwin", // intel macbooks
  "aarch64-apple-ios",
  "aarch64-apple-ios-simulator",
];

const _maxAttempts = 4;

bool _isTransient(Object error) {
  if (error is! DioException) {
    return false;
  }
  final status = error.response?.statusCode;
  if (status != null) {
    return status >= 500;
  }
  switch (error.type) {
    case DioExceptionType.connectionTimeout:
    case DioExceptionType.sendTimeout:
    case DioExceptionType.receiveTimeout:
    case DioExceptionType.connectionError:
      return true;
    default:
      return false;
  }
}

Future<T> _withRetry<T>(String what, Future<T> Function() operation) async {
  for (var attempt = 1;; attempt++) {
    try {
      return await operation();
    } catch (error) {
      if (attempt >= _maxAttempts || !_isTransient(error)) {
        rethrow;
      }
      final wait = Duration(seconds: 5 * attempt);
      print("  $what failed (attempt $attempt/$_maxAttempts): $error");
      print("  retrying in ${wait.inSeconds}s");
      await Future.delayed(wait);
    }
  }
}

Future<void> main() async {
  final resp = await _withRetry(
    "release list",
    () => _dio.get("https://api.github.com/repos/mrcyjanek/monero_c/releases"),
  );
  final data = resp.data[0];
  final tagName = data['tag_name'];
  print("Downloading artifacts for: ${tagName}");
  final assets = data['assets'] as List<dynamic>;
  final bundle = assets.firstWhere((asset) => asset["name"] == "release-bundle.zip");
  const bundlePath = "scripts/monero_c/release-bundle.zip";
  print("- downloading $bundlePath");
  await _withRetry(
    "release-bundle.zip",
    () => _dio.download(bundle["browser_download_url"] as String, bundlePath),
  );
  final archive = ZipDecoder().decodeStream(InputFileStream(bundlePath));
  for (final file in archive) {
    final parts = file.name.split("/");
    if (!file.isFile || parts.length != 3 || !triplets.contains(parts[1])) continue;
    final localFilename = "scripts/monero_c/release/${file.name}";
    print("  extracting $localFilename");
    Directory(File(localFilename).parent.path).createSync(recursive: true);
    final outputStream = OutputFileStream(localFilename);
    file.writeContent(outputStream);
    outputStream.closeSync();
  }
  File(bundlePath).deleteSync();
  if (Platform.isMacOS) {
    print("Generating ios framework");
    final result = Process.runSync(
      "bash",
      ["-c", "cd scripts/ios && ./gen_framework.sh && cd ../.."],
      environment: {"MONEROC_TAG": tagName as String},
    );
    print((result.stdout + result.stderr).toString().trim());
  }
}

File: tool/download_moneroc_prebuilds.py
#!/usr/bin/env python3

import os
import subprocess
import requests
import lzma
import shutil



# Define triplets list
triplets = [
    "x86_64-linux-gnu",
    "x86_64-linux-android",
    "aarch64-linux-android",
    "armv7a-linux-androideabi",
    # "x86_64-w64-mingw32",
    # "aarch64-apple-darwin",
    # "x86_64-apple-darwin",
    "aarch64-host-apple-darwin",
    # "aarch64-apple-ios",
    # "aarch64-apple-iossimulator",
]


def main():
    # Get the latest release data
    resp = requests.get("https://api.github.com/repos/mrcyjanek/monero_c/releases")
    data = resp.json()[0]
    tag_name = data["tag_name"]
    print(f"Downloading artifacts for: {tag_name}")

    assets = data["assets"]
    for asset in assets:
        for triplet in triplets:
            filename = asset["name"]
            if triplet not in filename:
                continue

            coin = filename.split("_")[0]
            local_filename = filename.replace(f"{coin}_{triplet}_", "")
            local_filename = (
                f"scripts/monero_c/release/{coin}/{triplet}_{local_filename}"
            )

            # Create directory if it doesn't exist
            os.makedirs(os.path.dirname(local_filename), exist_ok=True)

            url = asset["browser_download_url"]
            print(f"- downloading {local_filename}")

            # Download the file
            response = requests.get(url)
            with open(local_filename, "wb") as f:
                f.write(response.content)

            # Extract if it's an .xz file
            if local_filename.endswith(".xz"):
                print(f"  extracting {local_filename}")
                with lzma.open(local_filename) as f_in:
                    with open(local_filename.replace(".xz", ""), "wb") as f_out:
                        shutil.copyfileobj(f_in, f_out)

    # Generate iOS framework if on macOS
    if os.uname().sysname == "Darwin":  # Check if on macOS
        print("Generating ios framework")
        result = subprocess.run(
            ["bash", "-c", "cd scripts/ios && ./gen_framework.sh && cd ../.."],
            capture_output=True,
            text=True,
        )
        print(result.stdout.strip() + result.stderr.strip())


if __name__ == "__main__":
    main()
